# Rule Name: rule-id

Brief description of what this rule does.

## Rule Details

Detailed explanation of when this rule triggers and why it exists.

Examples of **incorrect** code for this rule:

```text
# Example of code that violates this rule
```

Examples of **correct** code for this rule:

```text
# Example of code that follows this rule
```

## Options

This rule has the following options:

- `option1`: Description of option 1
- `option2`: Description of option 2

Example configuration:

```json
{
  "rules": {
    "rule-id": ["error", {
      "option1": true
    }]
  }
}
```

## When Not To Use It

Guidance on scenarios where you might want to disable this rule.

## Related Rules

- [related-rule-1](./related-rule-1.md)
- [related-rule-2](./related-rule-2.md)

## Resources

- [Additional resource links if applicable]

## Version

Available since: 1.0.0

## Metadata

- **Category**: Category Name
- **Severity**: error | warning
- **Fixable**: Yes | No
